<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Model and Predict RBP Binding Preference &mdash; A little book about CLIP-seq 0.0.2 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/my_style.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="A little book about CLIP-seq 0.0.2 documentation" href="index.html" />
    <link rel="next" title="Pipelines, Toolkits and Softwares" href="c30pipeline.html" />
    <link rel="prev" title="Analyze CLIP-seq Data to locate RBP Binding Sites" href="c10anlyzRBP.html" />
 
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-49989911-1']);
  _gaq.push(['_trackPageview']);
</script>

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="c30pipeline.html" title="Pipelines, Toolkits and Softwares"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="c10anlyzRBP.html" title="Analyze CLIP-seq Data to locate RBP Binding Sites"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">A little book about CLIP-seq 0.0.2 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Model and Predict RBP Binding Preference</a><ul>
<li><a class="reference internal" href="#hidden-markov-model">Hidden Markov Model</a><ul>
<li><a class="reference internal" href="#sequence-path-v-s-state-path">Sequence Path v.s. State Path</a></li>
<li><a class="reference internal" href="#why-hidden">Why Hidden?</a></li>
<li><a class="reference internal" href="#think-in-hmm">Think in HMM</a><ul>
<li><a class="reference internal" href="#emission-probs">Emission Probs</a></li>
<li><a class="reference internal" href="#transition-probs">Transition Probs</a></li>
<li><a class="reference internal" href="#generate-sequence">Generate sequence</a></li>
</ul>
</li>
<li><a class="reference internal" href="#judge-possible-state-path">Judge Possible State Path</a><ul>
<li><a class="reference internal" href="#compute-probability-of-state-path">Compute Probability of State Path</a></li>
<li><a class="reference internal" href="#enumerate-state-paths">Enumerate State Paths</a></li>
<li><a class="reference internal" href="#viterbi-algorithm">Viterbi algorithm</a></li>
</ul>
</li>
<li><a class="reference internal" href="#estimate-hmm-in-action">Estimate HMM in action</a><ul>
<li><a class="reference internal" href="#supervised-learning">Supervised Learning</a></li>
<li><a class="reference internal" href="#unsupervised-learning">Unsupervised Learning</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#svm">SVM</a></li>
<li><a class="reference internal" href="#graph-kernel">Graph Kernel</a></li>
<li><a class="reference internal" href="#random-forest">Random Forest</a></li>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="c10anlyzRBP.html"
                        title="previous chapter">Analyze CLIP-seq Data to locate RBP Binding Sites</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="c30pipeline.html"
                        title="next chapter">Pipelines, Toolkits and Softwares</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/c20mdlPrdict.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="model-and-predict-rbp-binding-preference">
<h1>Model and Predict RBP Binding Preference<a class="headerlink" href="#model-and-predict-rbp-binding-preference" title="Permalink to this headline">¶</a></h1>
<p>With methods mentioned in previous chapter we could hopefully distinguish the informative signals from the sequencing noises. Results achieved at this step could be enough for most exploratory researches. However, what if we take a further step? What could we learn from the confident results? Could we learn about the binding preference of RNA binding proteins and construct models to unveil why RBP binds here but not there? Could we even predict the global binding sites of RBP? It does not necessarily means making CLIP-seq <em>in silico</em> possible, but machine learning methods could yield promising supplement to the real signals which could be left out by the empirical thresholding.</p>
<div class="section" id="hidden-markov-model">
<span id="c20-hmm"></span><h2>Hidden Markov Model<a class="headerlink" href="#hidden-markov-model" title="Permalink to this headline">¶</a></h2>
<p>In the beginning of this section I would like to introduce HMM by a toy application: 5&#8217; end splicing site recognition, which was originated from NBT paper <a class="reference internal" href="#eddy2004" id="id1">[Eddy2004]</a>. Admittedly this toy HMM is naive, it hopefully could help you take first step knowing how to infer the most possible splicing site, simply starting from raw DNA/RNA sequence together with some prior knowledges. For now we begin with the sequence: <em>CTTCATGTGAAAGCAGACGTAAGTCA</em>. Series of ideas and concepts would be illustrated step by step so that you could see how the biological question was translated to mathematical questions. Hopefully after you have known HMM I would introduce how HMM was employed in RBP science researches (<a class="reference internal" href="#zhang2013" id="id2">[Zhang2013]</a> and <a class="reference internal" href="#han2014" id="id3">[Han2014]</a>).</p>
<div class="section" id="sequence-path-v-s-state-path">
<h3>Sequence Path v.s. State Path<a class="headerlink" href="#sequence-path-v-s-state-path" title="Permalink to this headline">¶</a></h3>
<p>Sequence could be read by sequencing and thus observed. Each nucleotide residing on the DNA string belongs to some bio-functional regions. They are exons to be transcribed in the mature RNA, introns to be spliced out and splicing site, the boundary of exon and intron. These regions of different functions could be described as different states.</p>
<p>Four basic types of nucleotides (A, T, C, G) come together to form a sequence path. Likewise, the state path is composed of &#8220;E&#8221;, &#8220;I&#8221;, and &#8220;5&#8221;, denoting &#8220;Exon&#8221;, &#8220;Intron&#8221; and &#8220;5&#8217;-end Splicing Site&#8221; respectively. For convenience we disregard 3&#8217; end splicing sites here.</p>
<p>For example:</p>
<div class="highlight-python"><div class="highlight"><pre>Sequence Path: CTTCATGTGAAAGCAGACGTAAGTCA
State    Path: EEEEEEEEEEEEEEEEEE5IIIIIII
</pre></div>
</div>
<p>Therefore the ultimate goal of locating the splicing site is to figure out the state of each nucleotide and find the one under splicing site state. Generally speaking, given the sequence path we try to infer the state path.</p>
</div>
<div class="section" id="why-hidden">
<h3>Why Hidden?<a class="headerlink" href="#why-hidden" title="Permalink to this headline">¶</a></h3>
<p>At first sight whether GC-rich or AT-rich of a DNA string could be quickly determined given the sequence path, however you could barely say the state path immediately. This is where the &#8220;hidden&#8221; comes from.</p>
<p>The observed sequence path alone cannot help us to reveal the unobserved state path. With additional prior knowledge, for example, the distribution of A/T/G/C in each state, we could unveil the most possible E/I/5 state path in a <strong>top-bottom</strong> fashion.</p>
</div>
<div class="section" id="think-in-hmm">
<h3>Think in HMM<a class="headerlink" href="#think-in-hmm" title="Permalink to this headline">¶</a></h3>
<p>Suppose we have understood the unknown dark matter in the bottom and all the prior knowledge. Let&#8217;s think in HMM to generate a DNA sequence in <strong>bottom-up</strong> fashion. Oh, calm down, just playing roulette wheel.</p>
<div class="section" id="emission-probs">
<h4>Emission Probs<a class="headerlink" href="#emission-probs" title="Permalink to this headline">¶</a></h4>
<p>Let&#8217;s play with intron state and under intron state we set the bases distribution as:</p>
<div class="highlight-python"><div class="highlight"><pre>    A   C   G   T
I 0.4 0.1 0.1 0.4
</pre></div>
</div>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/c20HMM_Emission_Prob1.png"><img alt="_images/c20HMM_Emission_Prob1.png" src="_images/c20HMM_Emission_Prob1.png" /></a>
<p class="caption">Distribution of A, C, G, T in toy intron. <em>(See R scripts to generate pie chart in final section in case of your curiosity)</em></p>
</div>
<p>The <strong>emission probabilities</strong> under intron state could be described as the above roulette wheel. To generate a DNA string given the emission probabilities of intron, it is straightforward to implement sampling intron sequence of 26nt using <tt class="docutils literal"><span class="pre">R</span></tt> language:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> set.seed<span class="p">(</span><span class="m">1026</span><span class="p">)</span>
<span class="o">&gt;</span> dna.bases <span class="o">&lt;-</span> c<span class="p">(</span><span class="s">&quot;A&quot;</span><span class="p">,</span> <span class="s">&quot;C&quot;</span><span class="p">,</span> <span class="s">&quot;G&quot;</span><span class="p">,</span> <span class="s">&quot;T&quot;</span><span class="p">)</span>
<span class="o">&gt;</span> intron.emission.prob <span class="o">&lt;-</span> c<span class="p">(</span><span class="m">0.40</span><span class="p">,</span> <span class="m">0.10</span><span class="p">,</span> <span class="m">0.10</span><span class="p">,</span> <span class="m">0.40</span><span class="p">)</span>
<span class="o">&gt;</span> seq.len <span class="o">&lt;-</span> <span class="m">26</span>
<span class="o">&gt;</span> sample.seq <span class="o">&lt;-</span> sample<span class="p">(</span>x <span class="o">=</span> dna.bases<span class="p">,</span> prob <span class="o">=</span> intron.emission.prob<span class="p">,</span>
                                                size <span class="o">=</span> seq.len<span class="p">,</span> replace <span class="o">=</span> <span class="k-Variable">T</span><span class="p">)</span>
<span class="o">&gt;</span> cat<span class="p">(</span>sample.seq<span class="p">,</span> <span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="o">&gt;</span> table<span class="p">(</span>sample.seq<span class="p">)</span>
A C A G A A A A A A A <span class="k-Variable">T</span> G A <span class="k-Variable">T</span> A A A C A G G G A A <span class="k-Variable">T</span>
</pre></div>
</div>
<p>You could even double-check whether this string follow your prior set distribution. I give you result of 1000nt sequence by setting <tt class="docutils literal"><span class="pre">seq.len</span> <span class="pre">&lt;-</span> <span class="pre">1000</span></tt> without changing other codes:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> table<span class="p">(</span>sample.seq<span class="p">)</span>
sample.seq
  A   C   G   <span class="k-Variable">T</span>
<span class="m">396</span>  <span class="m">81</span> <span class="m">101</span> <span class="m">422</span>
</pre></div>
</div>
<p>Emission probabilities of each state is dedicated to generate observed base. Don&#8217;t forget we assume we are HMM now. If we clearly know state status of each base, we could immediately sample its nucleotide given the emission probabilities. Previously we set <tt class="docutils literal"><span class="pre">seq.len</span></tt> as <tt class="docutils literal"><span class="pre">26</span></tt> or <tt class="docutils literal"><span class="pre">1000</span></tt> to generate sequence of corresponding length. In the same way, we could give the individual base a nucleotide by setting <tt class="docutils literal"><span class="pre">seq.len</span> <span class="pre">&lt;-</span> <span class="pre">1</span></tt>.</p>
<p>This story under intron state could be exactly applied to exon, splicing sites states as well. Thus we have three wheels of emission probabilities under three different states.</p>
</div>
<div class="section" id="transition-probs">
<h4>Transition Probs<a class="headerlink" href="#transition-probs" title="Permalink to this headline">¶</a></h4>
<p>HMM has a feature: the next state is determined by ongoing state, or ongoing state is determined by previous one. For example, now we know we are under &#8220;Exon&#8221; state. When we considering generating next one, we are hesitating between whether continuing &#8220;Exon&#8221;, or changing to &#8220;Splicing Site&#8221; state. Moving from one state to another state is exactly determined by <strong>transition probabilities</strong>.</p>
<p>Though emission and transition probabilities have different biological missions: generating sequence and state path respectively. Their mathematical principle are exactly same - sampling something following multinomial model.</p>
<p>Let&#8217;s assume exon status has the following transition probabilities:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> state.bases <span class="o">&lt;-</span> c<span class="p">(</span><span class="s">&quot;E&quot;</span><span class="p">,</span> <span class="s">&quot;5&quot;</span><span class="p">,</span> <span class="s">&quot;I&quot;</span><span class="p">)</span>
<span class="o">&gt;</span> exon.transition.prob <span class="o">&lt;-</span> c<span class="p">(</span><span class="m">0.9</span><span class="p">,</span> <span class="m">0.1</span><span class="p">,</span> <span class="m">0</span><span class="p">)</span>
<span class="o">&gt;</span> exon.transition.matrix <span class="o">&lt;-</span> matrix<span class="p">(</span>exon.transition.prob<span class="p">,</span> nrow <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                                                                        byrow <span class="o">=</span> <span class="k-Variable">T</span><span class="p">)</span>
<span class="o">&gt;</span> colnames<span class="p">(</span>exon.transition.matrix<span class="p">)</span> <span class="o">&lt;-</span> state.bases
<span class="o">&gt;</span> row.names<span class="p">(</span>exon.transition.matrix<span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&quot;E&quot;</span>
<span class="o">&gt;</span> print<span class="p">(</span>exon.transition.matrix<span class="p">)</span>
    E   <span class="m">5</span> I
E <span class="m">0.9</span> <span class="m">0.1</span> <span class="m">0</span>
</pre></div>
</div>
<p>It means the probability of continuing exon state is 0.9 whereas changing to 5&#8217; end splicing site state is 0.1. Note that in our toy model exon cannot directly jump to intron state.</p>
<p>You cannot be unfamiliar with the following codes to choose state for next base:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> set.seed<span class="p">(</span><span class="m">1026</span><span class="p">)</span>
<span class="o">&gt;</span> sample.state <span class="o">&lt;-</span> sample<span class="p">(</span>x <span class="o">=</span> state.bases<span class="p">,</span> prob <span class="o">=</span> exon.transition.prob<span class="p">,</span>
                                                size <span class="o">=</span> <span class="m">1</span><span class="p">,</span> replace <span class="o">=</span> <span class="k-Variable">T</span><span class="p">)</span>
<span class="o">&gt;</span> cat<span class="p">(</span>sample.state<span class="p">)</span>
E
</pre></div>
</div>
</div>
<div class="section" id="generate-sequence">
<h4>Generate sequence<a class="headerlink" href="#generate-sequence" title="Permalink to this headline">¶</a></h4>
<p>So far hopefully you clear have learned:</p>
<ul class="simple">
<li>Generating sequence path purely under one state, i.e. intron</li>
<li>Determine state of next base when current base is under certain state, i.e. exon</li>
</ul>
<p>Next before generating sequences and of course state path we need make the emission and transition probabilities to the full story:</p>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/HMMandSplicingSite_Mx.png"><img alt="_images/HMMandSplicingSite_Mx.png" src="_images/HMMandSplicingSite_Mx.png" /></a>
<p class="caption">Diagram showing emission and transition probabilities matrix <a class="reference internal" href="#eddy2004" id="id4">[Eddy2004]</a>.</p>
</div>
<p>In particular there are additional &#8220;Start&#8221; and &#8220;End&#8221; states for model purpose. These two states have no base assignment and they serve as the <tt class="docutils literal"><span class="pre">^</span></tt> and <tt class="docutils literal"><span class="pre">$</span></tt> in the regular expression in programming. In the figure the transition probability from &#8220;Start&#8221; to &#8220;Exon&#8221; state is <tt class="docutils literal"><span class="pre">1.0</span></tt> which means in our toy example here we are going to generate sequence starting with &#8220;Exon&#8221;. Likewise the probability of continuing &#8220;Intron&#8221; is <tt class="docutils literal"><span class="pre">0.9</span></tt>.</p>
<p>The full emission probabilities matrix would be:</p>
<div class="highlight-bash"><div class="highlight"><pre>        A    C    G    T
S   0.00 0.00 0.00 0.00
E   0.25 0.25 0.25 0.25
SS5 0.05 0.00 0.95 0.00
I   0.40 0.10 0.10 0.40
N   0.00 0.00 0.00 0.00
</pre></div>
</div>
<p>The full transition probabilities matrix would be:</p>
<div class="highlight-bash"><div class="highlight"><pre>    S   E SS5   I   N
S   0 1.0 0.0 0.0 0.0
E   0 0.9 0.1 0.0 0.0
SS5 0 0.0 0.0 1.0 0.0
I   0 0.0 0.0 0.9 0.1
N   0 0.0 0.0 0.0 1.0
</pre></div>
</div>
<p>Check the source R script to find <tt class="docutils literal"><span class="pre">GenerateHMMSeq</span></tt> function. After you type the following in the R console, TA-DA, a sequence with HMM in mind is generated.</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> set.seed<span class="p">(</span><span class="m">1026</span><span class="p">)</span>
<span class="o">&gt;</span> GenerateHMMSeq<span class="p">(</span>emmisionMx<span class="o">=</span>myEmisMtx<span class="p">,</span> transitionMx<span class="o">=</span>myTransMtx<span class="p">,</span> len<span class="o">=</span><span class="m">20</span><span class="p">)</span>
Start
Base <span class="m">1</span> is A under E
Base <span class="m">2</span> is A under E
Base <span class="m">3</span> is C under E
Base <span class="m">4</span> is G under E
Base <span class="m">5</span> is G under E
Base <span class="m">6</span> is <span class="k-Variable">T</span> under E
Base <span class="m">7</span> is C under E
Base <span class="m">8</span> is C under E
Base <span class="m">9</span> is C under E
Base <span class="m">10</span> is G under SS5
Base <span class="m">11</span> is G under I
Base <span class="m">12</span> is A under I
Base <span class="m">13</span> is <span class="k-Variable">T</span> under I
Base <span class="m">14</span> is <span class="k-Variable">T</span> under I
Base <span class="m">15</span> is <span class="k-Variable">T</span> under I
End
Seq Path<span class="o">:</span> A A C G G <span class="k-Variable">T</span> C C C G G A <span class="k-Variable">T</span> <span class="k-Variable">T</span> <span class="k-Variable">T</span>
StatPath<span class="o">:</span> E E E E E E E E E SS5 I I I I I N
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">1</span>
</pre></div>
</div>
<p>You may noticed that in this case the length of output sequence is 15 though 20 is pre-defined. Is it wrong? No, because the probability of changing from &#8220;Intron&#8221; to &#8220;End&#8221; state is <tt class="docutils literal"><span class="pre">0.1</span></tt>, not <tt class="docutils literal"><span class="pre">0</span></tt>, Ongoing &#8220;Intron&#8221; state is possible to choose &#8220;End&#8221; state for next base.</p>
</div>
</div>
<div class="section" id="judge-possible-state-path">
<h3>Judge Possible State Path<a class="headerlink" href="#judge-possible-state-path" title="Permalink to this headline">¶</a></h3>
<p>So far hopefully you have known how to think in HMM to generate sequence path and state path given the transition and emission probabilities matrix available. Now back to our original goal: locating 5&#8217;-end splicing site. Now out mind goes in <strong>top-bottom</strong> direction and our mission is to find the state path with highest probabilities given the observed sequence path.</p>
<div class="section" id="compute-probability-of-state-path">
<h4>Compute Probability of State Path<a class="headerlink" href="#compute-probability-of-state-path" title="Permalink to this headline">¶</a></h4>
<p>Recall our example:</p>
<div class="highlight-bash"><div class="highlight"><pre>Sequence Path: CTTCATGTGAAAGCAGACGTAAGTCA
State    Path: EEEEEEEEEEEEEEEEEE5IIIIIII
</pre></div>
</div>
<p>Given the observed sequence of 26nt, how likely do you think is the state path among the hundreds of combinations? In general <span class="math">\({\mathbf P}\left(\mathbf{s},\pi \mid \text{HMM},\theta\right)\)</span> of an Hidden Marcov Model with parameters ( <span class="math">\(\theta\)</span> ) generates a state path ( <span class="math">\(\pi\)</span> ) and an observed sequence ( <span class="math">\(\mathbf{s}\)</span> ) , is the product of all the emission probabilities and transition probabilities that were used:</p>
<div class="math">
\[\left( 0.25^{18} * 0.95 * 0.4^3 * 0.1 * 0.1 * 0.4^2 \right) * \left(1 * 0.9^{17} * 0.1 * 1 * 0.9^6 * 0.1 \right)\]</div>
<p>26nt sequence has 26 emissions and 27 transitions (note the final end state).</p>
<p>See <tt class="docutils literal"><span class="pre">ComputeStatePathProb</span></tt> function for details and it could yield:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> demo.seq.path <span class="o">&lt;-</span> c<span class="p">(</span><span class="s">&quot;CTTCATGTGAAAGCAGACGTAAGTCA&quot;</span><span class="p">)</span>
<span class="o">&gt;</span> demo.state.path <span class="o">&lt;-</span> c<span class="p">(</span><span class="s">&quot;EEEEEEEEEEEEEEEEEE5IIIIIII&quot;</span><span class="p">)</span>
<span class="o">&gt;</span> ComputeStatePathProb<span class="p">(</span>seqPath <span class="o">=</span> demo.seq.path<span class="p">,</span> statePath <span class="o">=</span> demo.state.path<span class="p">,</span>
                     transMtx <span class="o">=</span> myTransMtx<span class="p">,</span> emisMtx <span class="o">=</span> myEmisMtx<span class="p">,</span> log <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">-41.21968</span>
</pre></div>
</div>
</div>
<div class="section" id="enumerate-state-paths">
<h4>Enumerate State Paths<a class="headerlink" href="#enumerate-state-paths" title="Permalink to this headline">¶</a></h4>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/HMMandSplicingSite.png"><img alt="_images/HMMandSplicingSite.png" src="_images/HMMandSplicingSite.png" /></a>
<p class="caption">Most possible state path.</p>
</div>
<p>Among all the possible combinations of state path, how to pick up the most possible state path? There is one naive way: enumerate all the candidates, compute their probability, and pick up the highest one.</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> InspectorHMM<span class="p">(</span>demo.seq.path<span class="p">,</span> myTransMtx<span class="p">,</span> myEmisMtx<span class="p">)</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;All State Path with non-zero Probability: &quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEE5IIIIIIIIIIIIIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">2.904208e-21</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEE5IIIIIIIIIIIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">8.621866e-20</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEE5IIIIIIIIIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">1.347167e-19</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEE5IIIIIIIIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">4.431469e-21</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEE5IIIIIIIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">2.769668e-21</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEE5IIIIIIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">1.731043e-21</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEE5IIIIIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">8.222452e-20</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEEEE5IIIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">6.761885e-21</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEEEEE5IIIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">3.211895e-19</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEEEEEE5IIIIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">1.056545e-20</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEEEEEEEE5IIIIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">1.254647e-18</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEEEEEEEEEE5IIIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">2.579454e-20</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEEEEEEEEEEE5IIII&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">1.612159e-20</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEEEEEEEEEEEE5III&quot;</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">7.657755e-19</span>
Best State Path and original Sequence
EEEEEEEEEEEEEEEEEE5IIIIIII
CTTCATGTGAAAGCAGACGTAAGTCA
Maximum Probability<span class="o">:</span> <span class="m">1.25464666093437e-18</span>
Maximum Prob <span class="p">(</span>log<span class="p">)</span><span class="o">:</span>  <span class="m">-41.2196776860225</span>
Posterior Decoding<span class="o">:</span>  <span class="m">0.461971754327234</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;EEEEEEEEEEEEEEEEEE5IIIIIII&quot;</span>
</pre></div>
</div>
<p>So far hopefully you have understood the principles of HMM. Obviously we cannot use the enumerate methods because when the length of sequence is increasing we could barely afford the time and computing resources to do that. We do have smarter algorithms. Keep reading.</p>
</div>
<div class="section" id="viterbi-algorithm">
<h4>Viterbi algorithm<a class="headerlink" href="#viterbi-algorithm" title="Permalink to this headline">¶</a></h4>
</div>
</div>
<div class="section" id="estimate-hmm-in-action">
<h3>Estimate HMM in action<a class="headerlink" href="#estimate-hmm-in-action" title="Permalink to this headline">¶</a></h3>
<p>In the 5&#8217;-end splicing end recognition toy model, the emission and transition probabilities are presumed so that we could estimate the most possible state path which is the best explanation for the observed sequence.</p>
<p>But the question is where we could know about the emission and transition probabilities of HMM before we tackle HMM? How could we learn the parameters of HMM? The solutions to this question would be split into two branches: supervised learnings and unsupervised learnings. The algorithms behind these would be introduced in the following. Luckily there are several convenient <tt class="docutils literal"><span class="pre">R</span></tt> packages that associate with HMM.</p>
<ul class="simple">
<li>HMM</li>
<li>RHmm</li>
<li>depmixS4</li>
<li>mhsmm</li>
</ul>
<div class="section" id="supervised-learning">
<h4>Supervised Learning<a class="headerlink" href="#supervised-learning" title="Permalink to this headline">¶</a></h4>
</div>
<div class="section" id="unsupervised-learning">
<h4>Unsupervised Learning<a class="headerlink" href="#unsupervised-learning" title="Permalink to this headline">¶</a></h4>
<p>To model PTB binding preference, two-state HMM was designed based on triplets in order to assess whether 3nt motif would segregate into two states and whether these two states had distinct emission probabilities <a class="reference internal" href="#han2014" id="id5">[Han2014]</a>. We could implement the codes to reproduce the results by using <tt class="docutils literal"><span class="pre">RHmm</span></tt> package with the following brief analysis pipeline:</p>
<ul class="simple">
<li>Raw data retrieve and munging;</li>
<li>Construct observed sequence path for each read;</li>
<li>Design two-state HMM;</li>
<li>Estimate emission and transition probabilities, the parameters (<span class="math">\(\theta\)</span> ) of HMM;</li>
<li>Visualization of results.</li>
</ul>
<p>If you are not interested in data cleaning which ironically takes up 90% of time of most data scientists, you could jump to HMM estimation.</p>
<p><a href="#id6"><span class="problematic" id="id7">:underline:`CLIP-seq Data Retrieve`</span></a>. PTB CLIP-seq sequencing reads were mapped to hg17 and adapted to hg18 in the latest <a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE19323">GEO data</a> . Author had kindly provided the genome coordinations of CLIP-seq peak clusters so that we could easily fetch the genome and transcript sequence of CLIP-seq clusters by using <tt class="docutils literal"><span class="pre">Bioconductor</span></tt>.</p>
<div class="highlight-r"><div class="highlight"><pre><span class="c1">## ---------------------</span>
<span class="c1">## PTB CLIP-seq data</span>
<span class="c1">## ---------------------</span>
<span class="c1">## ---------------------</span>
<span class="c1">## Data retrieve and processing</span>
<span class="c1">## ---------------------</span>

<span class="c1">## Genome fasta</span>
source<span class="p">(</span><span class="s">&quot;http://bioconductor.org/biocLite.R&quot;</span><span class="p">)</span>
biocLite<span class="p">(</span><span class="s">&quot;BSgenome.Hsapiens.UCSC.hg18&quot;</span><span class="p">)</span> 
<span class="c1"># ~800Mb, or alternatively download and install the souce pkg</span>
<span class="c1"># my.cmd &lt;- &quot;axel http://www.bioconductor.org/packages/release/data/annotation/src/contrib/BSgenome.Hsapiens.UCSC.hg18_1.3.19.tar.gz&quot; </span>
<span class="c1"># system(my.cmd)</span>
<span class="c1"># install.packages(&quot;BSgenome.Hsapiens.UCSC.hg18_1.3.19.tar.gz&quot;, repos = NULL, type = &quot;source&quot;)</span>
library<span class="p">(</span><span class="s">&quot;BSgenome.Hsapiens.UCSC.hg18&quot;</span><span class="p">)</span>
hg18.genome <span class="o">&lt;-</span> BSgenome.Hsapiens.UCSC.hg18

<span class="c1">## Peak bed corrdinates</span>
download.file<span class="p">(</span>url <span class="o">=</span> <span class="s">&quot;ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE19nnn/GSE19323/suppl/GSE19323_ptb_peak_cluster.bed.gz&quot;</span><span class="p">,</span> destfile <span class="o">=</span> <span class="s">&quot;ptb_peak_bed.gz&quot;</span><span class="p">)</span>
my.cmd <span class="o">&lt;-</span> <span class="s">&quot;gunzip ptb_peak_bed.gz&quot;</span> <span class="c1"># file.name would be ptb_peak_bed</span>
system<span class="p">(</span>my.cmd<span class="p">)</span>
ptb.peak <span class="o">&lt;-</span> read.table<span class="p">(</span>file <span class="o">=</span> <span class="s">&quot;ptb_peak_bed&quot;</span><span class="p">,</span> header <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
head<span class="p">(</span>ptb.peak<span class="p">)</span>
colnames<span class="p">(</span>ptb.peak<span class="p">)</span> <span class="o">&lt;-</span> c<span class="p">(</span><span class="s">&quot;Chr&quot;</span><span class="p">,</span> <span class="s">&quot;Start&quot;</span><span class="p">,</span> <span class="s">&quot;End&quot;</span><span class="p">,</span> <span class="s">&quot;Peak&quot;</span><span class="p">,</span> <span class="s">&quot;Height&quot;</span><span class="p">,</span> <span class="s">&quot;Strand&quot;</span><span class="p">)</span>
neg.indices <span class="o">&lt;-</span> which<span class="p">(</span>ptb.peak<span class="p">[,</span> <span class="m">6</span><span class="p">]</span> <span class="o">==</span> <span class="s">&quot;-&quot;</span><span class="p">)</span>

<span class="c1">## Fetch CLIP-seq Cluster Sequence</span>
library<span class="p">(</span><span class="s">&quot;GenomicRanges&quot;</span><span class="p">)</span>
ptb.peak.GRobj <span class="o">&lt;-</span> GRanges<span class="p">(</span>seqnames <span class="o">=</span> Rle<span class="p">(</span>as.character<span class="p">(</span>ptb.peak<span class="p">[,</span> <span class="m">1</span><span class="p">])),</span> 
                          ranges <span class="o">=</span> IRanges<span class="p">(</span>ptb.peak<span class="p">[,</span> <span class="m">2</span><span class="p">],</span> 
                                           ptb.peak<span class="p">[,</span> <span class="m">3</span><span class="p">],</span>
                                           names <span class="o">=</span> ptb.peak<span class="p">[,</span> <span class="m">4</span><span class="p">]),</span> 
                          strand <span class="o">=</span> Rle<span class="p">(</span>ptb.peak<span class="p">[,</span> <span class="m">6</span><span class="p">]))</span>
ptb.peak.seq <span class="o">&lt;-</span> getSeq<span class="p">(</span>hg18.genome<span class="p">,</span> seqnames<span class="p">(</span>ptb.peak.GRobj<span class="p">),</span> 
                       start<span class="p">(</span>ptb.peak.GRobj<span class="p">),</span> 
                       end<span class="p">(</span>ptb.peak.GRobj<span class="p">))</span> <span class="c1">## raw genome seq</span>
ptb.peak.seq<span class="p">[</span>neg.indices<span class="p">]</span> <span class="o">&lt;-</span> reverseComplement<span class="p">(</span>ptb.peak.rna<span class="p">[</span>neg.indices<span class="p">])</span> <span class="c1">## transcript sequence </span>
</pre></div>
</div>
<p><a href="#id8"><span class="problematic" id="id9">:underline:`Cook Observed Sequence Path`</span></a>. For each read the sequence is continuously string, for example <em>ACTGATCGG</em> string and they are not triplet-sequence: <em>ACT GAT CGG</em>. In addition the format of cluster sequence is still in DNA, not RNA. Therefore one more step is needed to cook data.</p>
<div class="highlight-r"><div class="highlight"><pre><span class="c1">## Cook to observations</span>
library<span class="p">(</span><span class="s">&quot;plyr&quot;</span><span class="p">)</span>
library<span class="p">(</span><span class="s">&quot;snowfall&quot;</span><span class="p">)</span>
ptb.peak.seq <span class="o">&lt;-</span> as.list<span class="p">(</span>as.character<span class="p">(</span>ptb.peak.seq<span class="p">))</span>
Seq2Obv <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>dataList<span class="p">,</span> size<span class="p">)</span> <span class="p">{</span>
    pass.indices <span class="o">&lt;-</span> sapply<span class="p">(</span>dataList<span class="p">,</span> nchar<span class="p">)</span>
    dataList <span class="o">&lt;-</span> dataList<span class="p">[</span>pass.indices <span class="o">&gt;=</span> size<span class="p">]</span>
    dataList <span class="o">&lt;-</span> llply<span class="p">(</span>dataList<span class="p">,</span> <span class="kr">function</span><span class="p">(</span>ele<span class="p">)</span> <span class="p">{</span>
        ele <span class="o">&lt;-</span> as.character<span class="p">(</span>dna2rna<span class="p">(</span>DNAString<span class="p">(</span>ele<span class="p">)))</span>
        start.i <span class="o">&lt;-</span> <span class="m">1</span>
        ele <span class="o">&lt;-</span> substring<span class="p">(</span>ele<span class="p">,</span> 
                         seq<span class="p">(</span>start.i<span class="p">,</span> nchar<span class="p">(</span>ele<span class="p">),</span> size<span class="p">),</span> 
                         seq<span class="p">((</span>start.i <span class="o">+</span> size <span class="o">-</span> start.i<span class="p">),</span> nchar<span class="p">(</span>ele<span class="p">),</span> size<span class="p">))</span>
        <span class="kr">return</span><span class="p">(</span>ele<span class="p">[</span>ele <span class="o">!=</span> <span class="s">&quot;&quot;</span><span class="p">])</span>
    <span class="p">})</span>
    dataList
<span class="p">}</span>

Seq2ObvSuper <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>dataList<span class="p">,</span> size<span class="p">,</span> startChar<span class="p">)</span> <span class="p">{</span>
    pass.indices <span class="o">&lt;-</span> sapply<span class="p">(</span>dataList<span class="p">,</span> nchar<span class="p">)</span>
    dataList <span class="o">&lt;-</span> dataList<span class="p">[</span>pass.indices <span class="o">&gt;=</span> <span class="p">(</span>size <span class="o">+</span> startChar <span class="o">-</span> <span class="m">1</span><span class="p">)]</span>
    dataList <span class="o">&lt;-</span> sfLapply<span class="p">(</span>dataList<span class="p">,</span> <span class="kr">function</span><span class="p">(</span>ele<span class="p">)</span> <span class="p">{</span>
        ele <span class="o">&lt;-</span> as.character<span class="p">(</span>dna2rna<span class="p">(</span>DNAString<span class="p">(</span>ele<span class="p">)))</span>
        start.i <span class="o">&lt;-</span> startChar
        ele <span class="o">&lt;-</span> substring<span class="p">(</span>ele<span class="p">,</span> 
                         seq<span class="p">(</span>start.i<span class="p">,</span> nchar<span class="p">(</span>ele<span class="p">),</span> size<span class="p">),</span> 
                         seq<span class="p">((</span>size <span class="o">+</span> start.i <span class="o">-</span> <span class="m">1</span><span class="p">),</span> nchar<span class="p">(</span>ele<span class="p">),</span> size<span class="p">))</span>
        <span class="kr">return</span><span class="p">(</span>ele<span class="p">[</span>ele <span class="o">!=</span> <span class="s">&quot;&quot;</span><span class="p">])</span>
    <span class="p">})</span>
    dataList
<span class="p">}</span>

ptb.peak.obv1 <span class="o">&lt;-</span> Seq2ObvSuper<span class="p">(</span>ptb.peak.seq<span class="p">,</span> size <span class="o">=</span> <span class="m">3</span><span class="p">,</span> startChar <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
ptb.peak.obv2 <span class="o">&lt;-</span> Seq2ObvSuper<span class="p">(</span>ptb.peak.seq<span class="p">,</span> size <span class="o">=</span> <span class="m">3</span><span class="p">,</span> startChar <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
ptb.peak.obv3 <span class="o">&lt;-</span> Seq2ObvSuper<span class="p">(</span>ptb.peak.seq<span class="p">,</span> size <span class="o">=</span> <span class="m">3</span><span class="p">,</span> startChar <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
ptb.peak.obvFull <span class="o">&lt;-</span> c<span class="p">(</span>ptb.peak.obv1<span class="p">,</span> ptb.peak.obv2<span class="p">,</span> ptb.peak.obv3<span class="p">)</span>
</pre></div>
</div>
<p><a href="#id10"><span class="problematic" id="id11">:underline:`Design HMM`</span></a>. Two-state HMM was designed: binding state or non-binding state.</p>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/c20HMM_PTB.png"><img alt="_images/c20HMM_PTB.png" src="_images/c20HMM_PTB.png" style="width: 600px;" /></a>
<p class="caption">PTBP1 protein binding model. A: Two-states HMM was designed and successfully trained. Transition probabilities was shown. B: Emission probabilities of the two states had different preference to triplets.</p>
</div>
<p>We are trying to reproduce it.</p>
<p><a href="#id12"><span class="problematic" id="id13">:underline:`Estimation`</span></a>.</p>
<div class="highlight-r"><div class="highlight"><pre><span class="c1">## -----------------------------</span>
<span class="c1">## Estimate HMM of PTB</span>
<span class="c1">## -----------------------------</span>
require<span class="p">(</span><span class="s">&quot;RHmm&quot;</span><span class="p">)</span>
ptb.HMM.estimated <span class="o">&lt;-</span> HMMFit<span class="p">(</span>ptb.peak.obvFull<span class="p">,</span> dis <span class="o">=</span> <span class="s">&quot;DISCRETE&quot;</span><span class="p">,</span> nStates <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
       control <span class="o">=</span> list<span class="p">(</span>init <span class="o">=</span> <span class="s">&quot;RANDOM&quot;</span><span class="p">))</span>
ptb.HMM.estimated.maxLLH <span class="o">&lt;-</span> ptb.HMM.estimated<span class="o">$</span>LLH
<span class="kr">for</span> <span class="p">(</span>i <span class="kr">in</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">)</span> <span class="p">{</span>
    ptb.HMM.estimated.each <span class="o">&lt;-</span> HMMFit<span class="p">(</span>ptb.peak.obvFull<span class="p">,</span> dis <span class="o">=</span> <span class="s">&quot;DISCRETE&quot;</span><span class="p">,</span> 
                                     nStates <span class="o">=</span> <span class="m">2</span><span class="p">,</span> 
                                     control <span class="o">=</span> list<span class="p">(</span>init <span class="o">=</span> <span class="s">&quot;RANDOM&quot;</span><span class="p">))</span>
    ptb.HMM.estimated.each.LLH <span class="o">&lt;-</span> ptb.HMM.estimated.each<span class="o">$</span>LLH
    cat<span class="p">(</span>ptb.HMM.estimated.maxLLH<span class="p">,</span> <span class="s">&quot;vs&quot;</span><span class="p">,</span> ptb.HMM.estimated.each.LLH<span class="p">,</span> <span class="s">&quot;\n&quot;</span><span class="p">)</span>
    <span class="kr">if</span> <span class="p">(</span>ptb.HMM.estimated.each.LLH <span class="o">&gt;</span> ptb.HMM.estimated.maxLLH<span class="p">)</span> <span class="p">{</span>
        ptb.HMM.estimated <span class="o">&lt;-</span> ptb.HMM.estimated.each
        ptb.HMM.estimated.maxLLH <span class="o">&lt;-</span> ptb.HMM.estimated.each.LLH
    <span class="p">}</span>
<span class="p">}</span>
print<span class="p">(</span>ptb.HMM.estimated<span class="p">)</span>

ptb.HMM.transMx <span class="o">&lt;-</span> ptb.HMM.estimated<span class="o">$</span>HMM<span class="o">$</span>transMat
ptb.HMM.emisMx  <span class="o">&lt;-</span> ptb.HMM.estimated<span class="o">$</span>HMM<span class="o">$</span>distribution<span class="o">$</span>proba
ptb.HMM.nStates <span class="o">&lt;-</span> ptb.HMM.estimated<span class="o">$</span>HMM<span class="o">$</span>distribution<span class="o">$</span>nStates
ptb.HMM.nSymbols <span class="o">&lt;-</span> ptb.HMM.estimated<span class="o">$</span>HMM<span class="o">$</span>distribution<span class="o">$</span>nLevels
ptb.HMM.symbols <span class="o">&lt;-</span> names<span class="p">(</span>ptb.HMM.emisMx<span class="p">[[</span>ptb.HMM.nStates<span class="p">]])</span>
ptb.HMM.emisMx  <span class="o">&lt;-</span> matrix<span class="p">(</span>unlist<span class="p">(</span>ptb.HMM.emisMx<span class="p">),</span> byrow <span class="o">=</span> <span class="k-Variable">T</span><span class="p">,</span> 
                          ncol <span class="o">=</span> ptb.HMM.nSymbols<span class="p">,</span> 
                          nrow <span class="o">=</span> ptb.HMM.nStates<span class="p">)</span>
colnames<span class="p">(</span>ptb.HMM.emisMx<span class="p">)</span> <span class="o">&lt;-</span> ptb.HMM.symbols
row.names<span class="p">(</span>ptb.HMM.emisMx<span class="p">)</span> <span class="o">&lt;-</span> c<span class="p">(</span><span class="s">&quot;State 1&quot;</span><span class="p">,</span> <span class="s">&quot;State 2&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>As a result we have estimated the parameters of HMM modeling PTB binding preferences.</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> ptb.HMM.estimated<span class="o">$</span>HMM

Model<span class="o">:</span>
<span class="o">------</span>
<span class="m">2</span> states HMM with discrete distribution

Initial probabilities<span class="o">:</span>
       Pi <span class="m">1</span>      Pi <span class="m">2</span>
  <span class="m">0.3279376</span> <span class="m">0.6720624</span>

Transition matrix<span class="o">:</span>
          State <span class="m">1</span>   State <span class="m">2</span>
State <span class="m">1</span> <span class="m">0.8124988</span> <span class="m">0.1875012</span>
State <span class="m">2</span> <span class="m">0.7279489</span> <span class="m">0.2720511</span>

Conditionnal distribution parameters<span class="o">:</span>

Distribution parameters<span class="o">:</span>
                AAA        AAC         AAG         AAU         ACA         ACC
State <span class="m">1</span> <span class="m">0.002053805</span> <span class="m">0.00171227</span> <span class="m">0.002921558</span> <span class="m">0.004062008</span> <span class="m">0.001702238</span> <span class="m">0.004119598</span>
State <span class="m">2</span> <span class="m">0.026659666</span> <span class="m">0.02669045</span> <span class="m">0.018581166</span> <span class="m">0.029270826</span> <span class="m">0.024849193</span> <span class="m">0.019070079</span>
                 ACG         ACU         AGA         AGC         AGG
State <span class="m">1</span> <span class="m">0.0009545203</span> <span class="m">0.007847316</span> <span class="m">0.003571282</span> <span class="m">0.005141578</span> <span class="m">0.004968771</span>
State <span class="m">2</span> <span class="m">0.0042811116</span> <span class="m">0.049837123</span> <span class="m">0.016124514</span> <span class="m">0.014469545</span> <span class="m">0.015340024</span>
                AGU         AUA        AUC         AUG        AUU         CAA
State <span class="m">1</span> <span class="m">0.007489506</span> <span class="m">0.003924728</span> <span class="m">0.01476462</span> <span class="m">0.005052538</span> <span class="m">0.01347329</span> <span class="m">0.005563721</span>
State <span class="m">2</span> <span class="m">0.019660018</span> <span class="m">0.015210379</span> <span class="m">0.03463434</span> <span class="m">0.025846413</span> <span class="m">0.04669401</span> <span class="m">0.013701655</span>
                CAC        CAG        CAU         CCA        CCC          CCG
State <span class="m">1</span> <span class="m">0.007087994</span> <span class="m">0.01072144</span> <span class="m">0.01785467</span> <span class="m">0.013107710</span> <span class="m">0.01146904</span> <span class="m">0.0036611494</span>
State <span class="m">2</span> <span class="m">0.020376993</span> <span class="m">0.01442269</span> <span class="m">0.02483846</span> <span class="m">0.003294134</span> <span class="m">0.00473662</span> <span class="m">0.0002132302</span>
               CCU          CGA          CGC          CGG         CGU
State <span class="m">1</span> <span class="m">0.03117673</span> <span class="m">0.0013737549</span> <span class="m">0.0027200873</span> <span class="m">4.233660e-03</span> <span class="m">0.005146979</span>
State <span class="m">2</span> <span class="m">0.02154137</span> <span class="m">0.0007451607</span> <span class="m">0.0002868091</span> <span class="m">1.914924e-05</span> <span class="m">0.001748012</span>
                CUA        CUC         CUG        CUU         GAA         GAC
State <span class="m">1</span> <span class="m">0.016619599</span> <span class="m">0.04564824</span> <span class="m">0.044308856</span> <span class="m">0.06249475</span> <span class="m">0.005378598</span> <span class="m">0.003293537</span>
State <span class="m">2</span> <span class="m">0.008393343</span> <span class="m">0.01467510</span> <span class="m">0.007747943</span> <span class="m">0.01867869</span> <span class="m">0.011986162</span> <span class="m">0.011816205</span>
                GAG         GAU         GCA         GCC         GCG        GCU
State <span class="m">1</span> <span class="m">0.005374742</span> <span class="m">0.006428859</span> <span class="m">0.006176074</span> <span class="m">0.007123844</span> <span class="m">0.001613111</span> <span class="m">0.01686248</span>
State <span class="m">2</span> <span class="m">0.007391663</span> <span class="m">0.013159560</span> <span class="m">0.008755279</span> <span class="m">0.010752949</span> <span class="m">0.001284622</span> <span class="m">0.02543406</span>
                GGA         GGC         GGG         GGU         GUA        GUC
State <span class="m">1</span> <span class="m">0.007897306</span> <span class="m">0.008914374</span> <span class="m">0.008259496</span> <span class="m">0.021292694</span> <span class="m">0.008996625</span> <span class="m">0.02168497</span>
State <span class="m">2</span> <span class="m">0.003091445</span> <span class="m">0.003453484</span> <span class="m">0.002429911</span> <span class="m">0.001723835</span> <span class="m">0.003756899</span> <span class="m">0.02106414</span>
                GUG        GUU          UAA         UAC          UAG
State <span class="m">1</span> <span class="m">0.018862773</span> <span class="m">0.02229759</span> <span class="m">0.0008606254</span> <span class="m">0.001296163</span> <span class="m">0.0007859372</span>
State <span class="m">2</span> <span class="m">0.005962911</span> <span class="m">0.02446250</span> <span class="m">0.0347718127</span> <span class="m">0.042850016</span> <span class="m">0.0336709869</span>
                UAU         UCA         UCC         UCG        UCU        UGA
State <span class="m">1</span> <span class="m">0.009279143</span> <span class="m">0.029338790</span> <span class="m">0.033592568</span> <span class="m">0.006527562</span> <span class="m">0.09150535</span> <span class="m">0.01019676</span>
State <span class="m">2</span> <span class="m">0.053780320</span> <span class="m">0.004609686</span> <span class="m">0.008172366</span> <span class="m">0.000611805</span> <span class="m">0.02612267</span> <span class="m">0.02084916</span>
               UGC        UGG        UGU         UUA         UUC         UUG
State <span class="m">1</span> <span class="m">0.01901294</span> <span class="m">0.02669897</span> <span class="m">0.03650746</span> <span class="m">0.017918991</span> <span class="m">0.070980521</span> <span class="m">0.035614819</span>
State <span class="m">2</span> <span class="m">0.01816547</span> <span class="m">0.00453399</span> <span class="m">0.03746990</span> <span class="m">0.001830695</span> <span class="m">0.005125166</span> <span class="m">0.001564879</span>
                UUU
State <span class="m">1</span> <span class="m">0.072478344</span>
State <span class="m">2</span> <span class="m">0.006707263</span>
</pre></div>
</div>
<p><a href="#id14"><span class="problematic" id="id15">:underline:`Visualization`</span></a>.</p>
<div class="highlight-r"><div class="highlight"><pre><span class="c1">## -----------------------------</span>
<span class="c1">## Visualize PTB HMM</span>
<span class="c1">## -----------------------------</span>
ptb.HMM.emisMx.order <span class="o">&lt;-</span> ptb.HMM.emisMx<span class="p">[,</span> order<span class="p">(</span>ptb.HMM.emisMx<span class="p">[</span><span class="m">1</span><span class="p">,</span> <span class="p">])]</span>
library<span class="p">(</span>reshape2<span class="p">)</span>
ptb.HMM.emisMx.order <span class="o">&lt;-</span> as.data.frame<span class="p">(</span>ptb.HMM.emisMx.order<span class="p">)</span>
ptb.HMM.emisMx.order<span class="o">$</span>State <span class="o">&lt;-</span> row.names<span class="p">(</span>ptb.HMM.emisMx.order<span class="p">)</span>
ptb.HMM.emisMx.order.melt <span class="o">&lt;-</span> melt<span class="p">(</span>ptb.HMM.emisMx.order<span class="p">,</span> id.vars <span class="o">=</span> <span class="s">&quot;State&quot;</span><span class="p">)</span>

p <span class="o">&lt;-</span> ggplot<span class="p">(</span>data <span class="o">=</span> ptb.HMM.emisMx.order.melt<span class="p">,</span> 
       aes<span class="p">(</span>x <span class="o">=</span> variable<span class="p">,</span> y <span class="o">=</span> value<span class="p">,</span> fill <span class="o">=</span> State<span class="p">))</span> <span class="o">+</span> 
    geom_bar<span class="p">(</span>stat<span class="o">=</span><span class="s">&quot;identity&quot;</span><span class="p">,</span> position<span class="o">=</span>position_dodge<span class="p">())</span> <span class="o">+</span>
    theme_bw<span class="p">()</span> <span class="o">+</span> 
    theme<span class="p">(</span>text <span class="o">=</span> element_text<span class="p">(</span>family <span class="o">=</span> <span class="s">&quot;xkcd&quot;</span><span class="p">,</span> size <span class="o">=</span> <span class="m">20</span><span class="p">),</span> 
          axis.text.x <span class="o">=</span> element_text<span class="p">(</span>angle <span class="o">=</span> <span class="m">90</span><span class="p">,</span> hjust <span class="o">=</span> <span class="m">1</span><span class="p">),</span> 
          legend.justification <span class="o">=</span> c<span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">),</span> 
          legend.position<span class="o">=</span>c<span class="p">(</span><span class="m">0.9</span><span class="p">,</span> <span class="m">0.7</span><span class="p">))</span> <span class="o">+</span> 
    xlab<span class="p">(</span><span class="s">&quot;Triplet&quot;</span><span class="p">)</span> <span class="o">+</span> ylab<span class="p">(</span><span class="s">&quot;Emission Prob&quot;</span><span class="p">)</span>  
</pre></div>
</div>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/c20_PTB_estimated_emis.png"><img alt="_images/c20_PTB_estimated_emis.png" src="_images/c20_PTB_estimated_emis.png" /></a>
</div>
</div>
</div>
</div>
<div class="section" id="svm">
<span id="c20-svm"></span><h2>SVM<a class="headerlink" href="#svm" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="graph-kernel">
<span id="c20-graphkernel"></span><h2>Graph Kernel<a class="headerlink" href="#graph-kernel" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="random-forest">
<span id="c20-randomforest"></span><h2>Random Forest<a class="headerlink" href="#random-forest" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<table class="docutils citation" frame="void" id="eddy2004" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Eddy2004]</td><td><em>(<a class="fn-backref" href="#id1">1</a>, <a class="fn-backref" href="#id4">2</a>)</em> Eddy, S. R. What is a hidden Markov model? Nat. Biotechnol. 22, 1315–1316 (2004).</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="zhang2013" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[Zhang2013]</a></td><td>Zhang, Chaolin, et al. &#8220;Prediction of clustered RNA-binding protein motif sites in the mammalian genome.&#8221; Nucleic acids research 41.14 (2013): 6793-6807.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="han2014" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Han2014]</td><td><em>(<a class="fn-backref" href="#id3">1</a>, <a class="fn-backref" href="#id5">2</a>)</em> Han, Areum, et al. &#8220;De Novo Prediction of PTBP1 Binding and Splicing Targets Reveals Unexpected Features of Its RNA Recognition and Function.&#8221; PLOS Computational Biology 10.1 (2014).</td></tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="c30pipeline.html" title="Pipelines, Toolkits and Softwares"
             >next</a> |</li>
        <li class="right" >
          <a href="c10anlyzRBP.html" title="Analyze CLIP-seq Data to locate RBP Binding Sites"
             >previous</a> |</li>
        <li><a href="index.html">A little book about CLIP-seq 0.0.2 documentation</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2014, Yun Yan.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.1.
    </div>
<div class="footer">This page uses <a href="http://analytics.google.com/">
Google Analytics</a> to collect statistics. You can disable it by blocking
the JavaScript coming from www.google-analytics.com.
<script type="text/javascript">
  (function() {
    var ga = document.createElement('script');
    ga.src = ('https:' == document.location.protocol ?
              'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    ga.setAttribute('async', 'true');
    document.documentElement.firstChild.appendChild(ga);
  })();
</script>
</div>

  </body>
</html>